ls1intum/Artemis
# `Assessment`: Fix score view for accepted complaints

### Checklist
#### General
#### Client

### Motivation and Context
I recently noticed that after accepting a complaint, the scores page would not display the updated score of the student, but instead would stay on the first result before the complaint got accepted. This PR fixes that.

### Description
The wrong complaint score was introduced by #6410. While in the exam mode, all results (of all correction rounds are needed to allow the instructor to navigate into on of the specific results (of one selected correction round). This however leads to the server response now containing two elements in case of complaints: `[result before complaint, result after complaint]`. The server returns that as an unordered Set, the client then sorts the list to have the earlier results first in the array. The latest result (that should be displayed on the scores page) therefore is last in the array, and now `last()` gets used to determine the result to show.

When working with complaints, the "View the Complaint" button also was not working, it opened the result before accepting the complaint. This is now also fixed by incrementing the correction round when necessary.

We also now hide the "start new assessment" button for programming practice mode participations, here manual assessments should not be possible.

To simplify the code, these buttons got extracted into their own component.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with manual assessment enabled
- 1 Exercise of any other type

Open the scores page and ensure that
- starting a new assessment works
- canceling and continuing an ongoing assessment works
- viewing accepted and rejected complaints works (for accepted complaints, the new result is shown)
- no "start assessment" button is shown for practice mode submissions.

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a Programming Exercise with manual assessment and 2 correction rounds, also an exercise of any other type.


Open the scores page and ensure that
- starting a new assessment works for correction rounds 1 and 2
- canceling and continuing an ongoing assessment works for both correction rounds
- viewing accepted and rejected complaints works (for accepted complaints, the new result is shown)
(see also testing steps of #6410 and issue #6395)

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress

#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage

#### Client

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exercise-scores.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5717/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-scores/exercise-scores.component.ts.html) | 96.12% | ✅  |
| [manage-assessment-buttons.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5717/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-scores/manage-assessment-buttons.component.ts.html) | 66.66% | ✅  |
